### PR TITLE
Fix flaky cancellation test

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
The e2e test to check that cancellations work was [flaky](https://github.com/cowprotocol/services/actions/runs/10696856101/job/29652879019) for a while now. It's possible that this happens more frequently after we merged the runloop refactor but IIUC the underlying issue must have been present (and known) forever.

The test previously asserted that the **last** order event after a cancellation was the cancellation itself.
However, if an order makes it into the auction before getting cancelled it will make it through the runloop (where we store the `ready` event) before we notice that the order was cancelled when building the **next** auction:
1. order gets put into auction
2. order gets cancelled (store `cancellation` event)
3. order progresses through runloop (store `ready` event)
4. finish runloop
5. build next auction **without** cancelled order

IMO this bug actually highlights quite nicely that off-chain cancellations don't give any hard guarantees which we even point out in the [openapi](https://github.com/cowprotocol/services/blob/main/crates/orderbook/openapi.yml#L64-L68) docs.

# Changes
Only assert that we actually have a cancellation order event but it does not have to be the last event.

## How to test
Ran flaky test in a loop on my machine to see if it's still flaky but could no longer reproduce the failure.

### CODE REVIEW

**Review by squadgazzz — APPROVED** [review]
🙌

**Review by sunce86 — COMMENTED** [review]
I think line 256 needs to be removed, since it can also fail sometimes.

As well as line 247 while we are at it.

I am ok with merging the PR because making sure full CI is ran on each PR is very important. 
But although the test is fixed, we still have the issue with `get_order_status` that this test originally tested as well.
I think we should add a comment (or create a gh issue) to revert this change and use `get_order_status` again, when the test is switched to run_loop_mode::SyncToBlockchain mode, since AFAIU, the root cause of the issue explained shouldn't happen in that case?

**Comment by sunce86** [comment]
> IMO this bug actually highlights quite nicely that off-chain cancellations don't give any hard guarantees which we even point out in the [openapi](https://github.com/cowprotocol/services/blob/main/crates/orderbook/openapi.yml#L64-L68) docs.

I don't think this is relevant to this issue. This issue is not about inflight orders. Indeed, we can't prevent orders being filled once they are inflight. But this issue is about us running the next Auction with cancelled order even when our database is aware that it is cancelled, just because the Auction cut is done in a background task. So it's definitely up to us and can/should be fixed with SyncToBlockchain.

**Review by m-lord-renkse — APPROVED** [review]

**Comment by MartinquaXD** [comment]
> I think line 256 needs to be removed, since it can also fail sometimes.

I don't think that's correct. The test does not configure any drivers so the order can't ever be settled so the order status returned from the orderbook is not susceptible to race conditions in this test.

> I think we should add a comment (or create a gh issue) to revert this change and use get_order_status again, when the test is switched to run_loop_mode::SyncToBlockchain mode, since AFAIU, the root cause of the issue explained shouldn't happen in that case?

Seems reasonable. Although I'm not super sure we should make the test rely on these implementation specific things. 🤷‍♂️  

> I don't think this is relevant to this issue.

Why not? Of course it would have been possible to bend over backwards in the old runloop mode to remove solutions for cancelled orders but we didn't because that would have been unreasonable complexity for a rare race condition. The basic underlying condition that the effectiveness of off-chain cancellations 100% depend on the implementation of the off-chain components still stands.

**Comment by sunce86** [comment]
> I don't think that's correct. The test does not configure any drivers so the order can't ever be settled so the order status returned from the orderbook is not susceptible to race conditions in this test.

For the same reason you switched line 217? The test actually still fails locally:
Screenshot 2024-09-05 at 09 43 20

> Why not?

I meant in a sense that docs mention inflight_orders as the only reason, which is not the case here. 

> because that would have been unreasonable complexity for a rare race condition

Agree that this is not a proper way to solve it, but it's not that rare race condition? I mean, Auction refresh is every 10s, it takes 2s to update, so whenever a user cancels the order 12s or less before the end of runloop iteration, the next runloop iteration will overwrite database event store label to `Ready` which is not supposed to happen for canceled orders.

**Comment by MartinquaXD** [comment]
> For the same reason you switched line 217? The test actually still fails locally:

Sorry, was too trigger happy. For some reason I thought the line in question fetches the full order and asserts on the state of that instead of using the status endpoint where this is indeed a problem.
Out of curiosity how often did you have to run the test for it to fail. I think yesterday I let it run >30 times and never saw the issue.
Will open another PR to resolve this.

**Comment by sunce86** [comment]
> Out of curiosity how often did you have to run the test for it to fail.

For some reason it fails right away on my machine 😄

File: crates/e2e/tests/e2e/order_cancellation.rs [hunk 1]
```diff
@@ -214,10 +214,10 @@ async fn order_cancellation(web3: Web3) {
             .status,
         OrderStatus::Cancelled,
     );
-    assert_eq!(
-        services.get_order_status(&order_uids[0]).await.unwrap(),
-        orderbook::dto::order::Status::Cancelled,
-    );
+    let events = crate::database::events_of_order(services.db(), &order_uids[0]).await;
+    assert!(events
```
Comment (fleupold) [inline crates/e2e/tests/e2e/order_cancellation.rs:218]: nit: maybe comment on the counterintuitive use of `any` (due to the soft nature of offchain cancellations, it's possible that an order gets scheduled, ie ready, after being cancelled)

**Pull request merged**